### PR TITLE
8311902: Concurrency regression in the PBKDF2 key impl of SunJCE provider

### PR DESCRIPTION
**Pull request by valeriepeng**
This change adds back the Reference.ReachabilityFence(Object) call removed by [JDK-8301553](https://bugs.openjdk.org/browse/JDK-8301553).

Please help review.
Thanks!
Valerie

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311902](https://bugs.openjdk.org/browse/JDK-8311902): Concurrency regression in the PBKDF2 key impl of SunJCE provider (**Bug** - P2)

### Reviewers
 * [Anthony Scarpino](https://openjdk.org/census#ascarpino) (@ascarpino - **Reviewer**) ⚠️ Review applies to [9d3c15a7](https://git.openjdk.org/jdk/pull/14859/files/9d3c15a77afeac384c1b9ffbe89cd9d38595679b)
 * [Xue-Lei Andrew Fan](https://openjdk.org/census#xuelei) (@XueleiFan - **Reviewer**)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14859/head:pull/14859` \
`$ git checkout pull/14859`

Update a local copy of the PR: \
`$ git checkout pull/14859` \
`$ git pull https://git.openjdk.org/jdk.git pull/14859/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14859`

View PR using the GUI difftool: \
`$ git pr show -t 14859`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14859.diff">https://git.openjdk.org/jdk/pull/14859.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14859#issuecomment-1633471813)

### CODE REVIEW

**Review by ascarpino — APPROVED** [review]
Looks good to me

**Review by XueleiFan — APPROVED** [review]
It looks good to me to rollback to previous behaviors.  I was just wondering, if the use of key in other methods, like hashCode()/equals(), has the similar issue?  Thanks!

File: src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java
Comment (seanjmullan) [inline src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java:1]: I believe you also want to revert the changes made to `getPassword()`.

Comment (seanjmullan) [inline src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java:1]: I also think the change which moved the registering of the `Cleaner` outside the `finally` block in the constructor is not correct, as the passwd is no longer zero-ed out if the code after that throws an Exception.

Comment (valeriepeng) [reply src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java:1]: Yes, I will do that.

Comment (valeriepeng) [reply src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java:1]: Per my reading of the code. the cleaner is only used when the PBKDF2 key constructor succeeds. If an exception occurred, then the passwd cleanup is handled by the if (key == null) condition in the finally block.

**Comment by valeriepeng** [comment]
> It looks good to me to rollback to previous behaviors. I was just wondering, if the use of key in other methods, like hashCode()/equals(), has the similar issue? Thanks!

For the usage of hashCode()/equals(), there should be strong reference to the key object for the usage scenarios I'd think. Thus, probably not an issue?

File: src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java
Comment (seanjmullan) [reply src/java.base/share/classes/com/sun/crypto/provider/PBKDF2KeyImpl.java:1]: Yes, took another closer look at the code and you are right. So, never mind this comment.

**Comment by XueleiFan** [comment]
> > It looks good to me to rollback to previous behaviors. I was just wondering, if the use of key in other methods, like hashCode()/equals(), has the similar issue? Thanks!
> 
> For the usage of hashCode()/equals(), there should be strong reference to the key object for the usage scenarios I'd think. Thus, probably not an issue?

Yes, it makes sense to me.

**Comment by XueleiFan** [comment]
> > > It looks good to me to rollback to previous behaviors. I was just wondering, if the use of key in other methods, like hashCode()/equals(), has the similar issue? Thanks!
> > 
> > 
> > For the usage of hashCode()/equals(), there should be strong reference to the key object for the usage scenarios I'd think. Thus, probably not an issue?
> 
> Yes, it makes sense to me.

I may reply too quickly to get the idea.  Why you think there will be strong reference to the key object for hashCode()/equals(), but not for getEncoded()?

**Comment by valeriepeng** [comment]
> > > > It looks good to me to rollback to previous behaviors. I was just wondering, if the use of key in other methods, like hashCode()/equals(), has the similar issue? Thanks!
> > > 
> > > 
> > > For the usage of hashCode()/equals(), there should be strong reference to the key object for the usage scenarios I'd think. Thus, probably not an issue?
> > 
> > 
> > Yes, it makes sense to me.
> 
> I may reply too quickly to get the idea. Why you think there will be strong reference to the key object for hashCode()/equals(), but not for getEncoded()? I did not get the idea.

hashCode()/equals() are normally used when objects are still around, e.g. used in hash map. As for getEncoded(), callers may just want to retrieve the bytes and not keep the Key object.
After some internal discussion, I am inclined to include all methods including the hashCode(), equals() into this reachabilityFence() change for the sake of consistency.

**Review by XueleiFan — APPROVED** [review]
I have no more comments.  Thanks!

**Review by seanjmullan — APPROVED** [review]
Looks good. I think you should file a follow-on issue to make similar changes to DESKey, DESedeKey, PBEKey and other security classes that use Cleaner.

**Comment by valeriepeng** [comment]
Yes, I do plan a follow-on issue to apply this to other security classes which use Cleaner...
Thanks all for the review and comments~
